service/service.go

```diff
@@ -170,59 +173,96 @@ func NewService(cfg *Config, datadir string) (*Service, error) {
+
+	roundResults := make(chan roundResult, 1)
+
+	// Resume recovered rounds
+	for _, round := range roundsToResume {
+		round := round
+		s.executingRounds[round.ID] = struct{}{}
+		end := s.roundEndTime(round)
+		eg.Go(func() error {
+			err := round.recoverExecution(ctx, round.stateCache.Execution, end)
+			roundResults <- roundResult{round: round, err: err}
+			return nil
+		})
+	}
+
+	timer := s.scheduleRound(ctx, s.openRound)
 
 	for {
-		s.openRoundMutex.RLock()
-		epoch := s.openRound.Epoch()
-		s.openRoundMutex.RUnlock()
-
-		start := s.genesis.Add(s.cfg.EpochDuration * time.Duration(epoch)).Add(s.cfg.PhaseShift)
-		waitTime := time.Until(start)
-		timer := time.After(waitTime)
-		if waitTime > 0 {
-			log.Info("Round %v waiting for execution to start for %v", s.openRoundID(), waitTime)
-		}
 		select {
+		case cmd := <-cmds:
+			cmd(s)
+
+		case result := <-roundResults:
+			if result.err == nil {
+				broadcaster := s.broadcaster
+				go broadcastProof(s, result.round, result.round.execution, broadcaster)
+			}
+			delete(s.executingRounds, result.round.ID)
+
 		case <-timer:
+			round := s.openRound
+			s.openRound = s.newRound(ctx, round.Epoch()+1)
+			s.executingRounds[round.ID] = struct{}{}
+
+			end := s.roundEndTime(round)
+			minMemoryLayer := s.minMemoryLayer
+			eg.Go(func() error {
+				roundResults <- executeRound(ctx, round, end, minMemoryLayer)
+				return nil
+			})
+
+			// schedule the next round
+			timer = s.scheduleRound(ctx, s.openRound)
+
 		case <-ctx.Done():
-			log.Info("service shutting down")
-			s.openRoundMutex.Lock()
-			s.openRound = nil
-			s.openRoundMutex.Unlock()
+			logger.Info("service shutting down")
 			return
 		}
+	}
+}
 
-		s.openRoundMutex.Lock()
-		prevRound := s.openRound
-		s.newRound(ctx, prevRound.Epoch()+1)
-		s.openRoundMutex.Unlock()
+func (s *Service) roundStartTime(round *round) time.Time {
+	return s.genesis.Add(s.cfg.PhaseShift).Add(s.cfg.EpochDuration * time.Duration(round.Epoch()))
+}
 
-		executingRounds.Go(func() error {
-			round := prevRound
-			if err := s.executeRound(ctx, round); err != nil {
-				s.asyncError(fmt.Errorf("round %v execution error: %v", round.ID, err))
-				return nil
-			}
-			broadcastProof(s, round, round.execution, s.getBroadcaster())
-			return nil
-		})
+func (s *Service) roundEndTime(round *round) time.Time {
+	return s.roundStartTime(round).Add(s.cfg.EpochDuration).Add(-s.cfg.CycleGap)
+}
+
+func (s *Service) scheduleRound(ctx context.Context, round *round) <-chan time.Time {
+	waitTime := time.Until(s.roundStartTime(round))
+	timer := time.After(waitTime)
+	if waitTime > 0 {
+		logging.FromContext(ctx).With().Info("waiting for execution to start", log.Duration("wait time", waitTime), log.String("round", round.ID))
 	}
+	return timer
 }
 
-func (s *Service) Start(b Broadcaster, atxProvider challenge_verifier.Verifier) error {
+func (s *Service) Start(b Broadcaster, verifier challenge_verifier.Verifier) error {
 	s.Lock()
 	defer s.Unlock()
```

Comment:
can get rid of the mutex using sync.Once?

Reply:
sure. up to you. it's used often in go-spacemesh. 

> It would be explicit that Service cannot be restarted after Shutdown()

i don't see there is any way to restart it after the Shutdown() RPC call. is there?

Reply:
No, there isn't even a shutdown RPC. `Shutdown()` is called on program exit.